File: src/action_tutorials_cpp/src/fibonacci_action_server.cpp
#include <chrono>
#include <memory>
#include <vector>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include "action_tutorials_interfaces/action/fibonacci.hpp"

using namespace std::chrono_literals;
using Fibonacci = action_tutorials_interfaces::action::Fibonacci;

class FibonacciActionServer : public rclcpp::Node {
public:
    FibonacciActionServer(int timeout_seconds, double sleep_seconds)
        : Node("fibonacci_action_server"),
          timeout_seconds_(timeout_seconds),
          sleep_duration_(std::chrono::duration<double>(sleep_seconds)),
          start_time_(this->get_clock()->now()) {
                    
        action_server_ = rclcpp_action::create_server<Fibonacci>(
            this,
            "fibonacci",
            std::bind(&FibonacciActionServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
            std::bind(&FibonacciActionServer::handle_cancel, this, std::placeholders::_1),
            std::bind(&FibonacciActionServer::handle_accepted, this, std::placeholders::_1)
        );

        timer_ = this->create_wall_timer(
            std::chrono::seconds(timeout_seconds_),
            std::bind(&FibonacciActionServer::shutdown_node, this)
        );
    }

private:
    rclcpp_action::Server<Fibonacci>::SharedPtr action_server_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Time start_time_;
    int timeout_seconds_;
    std::chrono::duration<double> sleep_duration_;

    rclcpp_action::GoalResponse handle_goal(
        const rclcpp_action::GoalUUID & uuid,
        std::shared_ptr<const Fibonacci::Goal> goal) {
        RCLCPP_INFO(this->get_logger(), "Received goal request with order %d", goal->order);
        (void)uuid;
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>> goal_handle) {
        RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>> goal_handle) {
        std::thread([this, goal_handle]() { execute(goal_handle); }).detach();
    }

    void execute(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>> goal_handle) {
        RCLCPP_INFO(this->get_logger(), "Executing goal...");

        const auto goal = goal_handle->get_goal();
        auto feedback = std::make_shared<Fibonacci::Feedback>();
        auto & sequence = feedback->partial_sequence;
        sequence.push_back(0);
        sequence.push_back(1);

        for (int i = 1; i < goal->order; ++i) {
            // if (goal_handle->is_canceling()) {
            //     goal_handle->canceled(result);
            //     RCLCPP_INFO(this->get_logger(), "Goal canceled");
            //     return;
            // }

            sequence.push_back(sequence[i] + sequence[i - 1]);
            goal_handle->publish_feedback(feedback);
            RCLCPP_INFO(this->get_logger(), "Feedback: [%d]", sequence.back());

            std::this_thread::sleep_for(sleep_duration_);
        }

        auto result = std::make_shared<Fibonacci::Result>();

        result->sequence = sequence;
        goal_handle->succeed(result);
        RCLCPP_INFO(this->get_logger(), "Goal succeeded");
    }

    void shutdown_node() {
        RCLCPP_INFO(this->get_logger(), "Timeout reached, shutting down node...");
        rclcpp::shutdown();
    }
};

int main(int argc, char ** argv) {

    rclcpp::init(argc, argv);

    if (argc != 3)
    {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Usage: ros2 run your_package your_server_node <timeout_seconds> <sleep_seconds>");
      return 1;
    }

    int timeout_seconds = std::stoi(argv[1]);
    double sleep_seconds = std::stod(argv[2]);

    auto server = std::make_shared<FibonacciActionServer>(timeout_seconds, sleep_seconds);
    rclcpp::spin(server);

    rclcpp::shutdown();
    return 0;
}

File: src/action_tutorials_py/action_tutorials_py/fibonacci_action_server.py
import sys
import time
import rclpy
from rclpy.action import ActionServer
from rclpy.node import Node
from action_tutorials_interfaces.action import Fibonacci


class FibonacciActionServer(Node):

    def __init__(self, timeout_seconds, sleep_seconds):
        super().__init__('fibonacci_action_server')

        self.sleep_duration = sleep_seconds
        self.timeout = timeout_seconds

        self._action_server = ActionServer(
            self,
            Fibonacci,
            'fibonacci',
            self.execute_callback
        )

        self.start_time = self.get_clock().now()
        self.create_timer(self.timeout, self.shutdown_node)

    def execute_callback(self, goal_handle):
        self.get_logger().info('Executing goal...')

        feedback_msg = Fibonacci.Feedback()
        feedback_msg.partial_sequence = [0, 1]

        # Move somewhere else
        if (self.get_clock().now() - self.start_time).nanoseconds / 1e9 >= self.timeout:
            self.get_logger().info('Timeout reached. Shutting down...')
            sys.exit(0)

        for i in range(1, goal_handle.request.order):
            if goal_handle.is_cancel_requested:
                self.get_logger().info('Goal canceled.')
                goal_handle.canceled()
                return Fibonacci.Result()

            feedback_msg.partial_sequence.append(
                feedback_msg.partial_sequence[i] + feedback_msg.partial_sequence[i - 1]
            )
            self.get_logger().info(f'Feedback: {feedback_msg.partial_sequence}')

            goal_handle.publish_feedback(feedback_msg)

            time.sleep(self.sleep_duration)

        goal_handle.succeed()
        result = Fibonacci.Result()
        result.sequence = feedback_msg.partial_sequence
        return result

    def shutdown_node(self):
        """Encerra o nó ao atingir o timeout."""
        self.get_logger().info('Shutting down node...')
        sys.exit(0)


def main(args=None):
    rclpy.init()

    if len(sys.argv) != 3:
        print("Usage: ros2 run your_package your_server_node <timeout_seconds> <sleep_seconds>")
        return

    try:
        timeout_seconds = int(sys.argv[1])
        sleep_seconds = float(sys.argv[2])
    except ValueError:
        print("Please provide valid integer for timeout and float for sleep.")
        return

    fibonacci_action_server = FibonacciActionServer(timeout_seconds, sleep_seconds)

    rclpy.spin(fibonacci_action_server)

    fibonacci_action_server.destroy_node()

    rclpy.shutdown()


if __name__ == '__main__':
    main()
